Git diff parsing

jakethekoenig opened the pull request:
A parser for the git diff format is introduced. And a test for the translation script between block, replacement and git_diff formats is introduced.

jakethekoenig commented on tests/parser_tests/translate_scripts_test.py:
```diff
@@ -0,0 +1,60 @@
+        return f.read()
+
+
+# This one doesn't pass. I think it's actually a bug with the parser as if I print out the final parsed
```
@PCSwingle I'll look into this tomorrow but I wanted to bring your attention to it because I thought you might just know what the problem is.

PCSwingle replied:
I'm looking into it. You can also mark tests like this with @pytest.mark.xfail to mark it as failing but still let everything pass

PCSwingle replied:
Holy cow I found the bug and it's actually a massive bug; I can't believe this isn't caught by any of our tests. I'll push a commit soon that fixes it (super easy fix) and will make it obvious what the bug was haha. Thanks for finding this!!

jakethekoenig commented:
I also have no idea why the `code_context` test is failing. I guess it's probably because I changed `testbed`.

PCSwingle commented on scripts/translate_transcript.py:
```diff
@@ -2,65 +2,78 @@
     "replacement": ReplacementParser(),
     "split-diff": SplitDiffParser(),
     "unified-diff": UnifiedDiffParser(),
+    "git": GitParser(),
```
I don't really think it should be added to this map; it isn't a parser, we don't have a prompt for it, and the user can't select it

PCSwingle replied:
Oh whoops, somehow I completely missed what file this was in! Ignore this haha

PCSwingle commented on testbed/format_examples/git_diff.txt:
Are you going to add any example of creating/deleting files (it would have a /dev/null instead of a/a or b/a)

jakethekoenig replied:
I believe it has both. See later in the file. I think moves, creates and deleted are all correctly supported

PCSwingle replied:
Ok I must be blind haha, I could've sworn I didn't see any /dev/null

PCSwingle commented on scripts/translate_transcript.py:
```diff
@@ -2,65 +2,78 @@
-ending_parser = parser_map[args.ending_format]
+def translate_message(message: str, starting_parser, ending_parser) -> str:
+    if hasattr(starting_parser, "parse_string"):
+        parsedLLMResponse = starting_parser.parse_string(message)
```
We could get rid of this hasattr if we made a stream_and_parse_llm_response that just collected the async generator back into a string and passed it to parse_string; just an idea though, if you don't like it no need to do it (it is a bit weird converting a string to a generator and back haha

jakethekoenig replied:
That's probably a good idea. I'll do that. Could also see giving parsers a parse_string method since it might be useful in other contexts and makes more sense

PCSwingle replied:
Maybe, but I think that would take a lot of work making a parse_string method for all of the parsers, since it's so interlinked with streaming right now.

jakethekoenig replied:
Yeah, you're right. I did it your way.

jakethekoenig commented on mentat/parsers/parser.py:
```diff
@@ -235,6 +235,7 @@ async def stream_and_parse_llm_response(
                                 cur_file_edit.rename_file_path = (
                                     file_edit.rename_file_path
                                 )
+                            cur_file_edit.replacements.extend(file_edit.replacements)
```
oof. Thanks for fixing!

jakethekoenig commented:
> I also have no idea why the `code_context` test is failing. I guess it's probably because I changed `testbed`.

I went ahead and just changed the test. I think any time any file in testbed is edited the exact token counts should change. I wonder if we should fix the test to a git commit to prevent that.

jakethekoenig commented:
I tried to get gpt-4-32k (because I wanted to fit parser.py into context) to write this for me. It didn't work the first try and I actually implementing all the parser methods was more than we needed for this pr so I decided to do it myself. Interestingly it first wrote out some methods and added todos which is why I opened https://github.com/AbanteAI/mentat/issues/189 and https://github.com/AbanteAI/mentat/issues/188
[transcript_20231019_073927.log](https://github.com/AbanteAI/mentat/files/13056078/transcript_20231019_073927.log)

PCSwingle commented:
> > I also have no idea why the `code_context` test is failing. I guess it's probably because I changed `testbed`.
> 
> I went ahead and just changed the test. I think any time any file in testbed is edited the exact token counts should change. I wonder if we should fix the test to a git commit to prevent that.

Yeah, that test is definitely pretty brittle; I definitely think it would be good to completely remove the testbed in the test and run it on only like 2 files (but we don't need to do that in this pr)

PCSwingle reviewed (approved):
Looks good to me!

biobootloader commented:
I added an issue for improving the brittle CodeContext test: https://github.com/AbanteAI/mentat/issues/201

Nice work!
